design/storage-class-device-set.md

```diff
@@ -0,0 +1,487 @@
+                  operator: In
+                  values:
+                    - cluster1
+                topologyKey: "failure-domain.beta.kubernetes.io/zone"
+      volumeClaimTemplates:
+      - spec:
+          resources:
+            requests:
+              storage: 5Ti
+          storageClassName: gp2-ebs
+    - name: cluster1-set2
+      count: 3
+      resources:
+        requests:
+          cpu: 2
+          memory: 8Gi
+      placement:
+        podAntiAffinity:
+          preferredDuringSchedulingIgnoredDuringExecution:
+          - weight: 100
+            podAffinityTerm:
+              labelSelector:
+                matchExpressions:
+                - key: "rook.io/cluster"
+                  operator: In
+                  values:
+                    - cluster1
+                topologyKey: "failure-domain.beta.kubernetes.io/zone"
+      volumeClaimTemplates:
+      - spec:
+          resources:
+            requests:
+              storage: 10Ti
+          storageClassName: gp2-ebs
+```
+
+In this example, we have two `StorageClassDeviceSets` with different capacities
+for the devices in each set. The devices with larger capacities would require a
+greater amount of memory to operate the OSD Pods, so that is reflected in the
+`resources` field.
+
+#### Example 4: Simple local storage
+
+```yaml
+type: CephCluster
+name: cluster1
+...
+spec:
+  ...
+  storage:
+    ...
+    nodes:
+    - name: node1
+    - name: node2
+    - name: node3
+    storageClassDeviceSet:
+    - name: cluster1-set1
+      count: 3
+      volumeClaimTemplates:
+      - spec:
+          resources:
+            requests:
+              storage: 1
+          storageClassName: cluster1-local-storage
+```
+
+In this example, we've defined that we already have three nodes. The nodes would
+be configured with one local storage device each, but they would not be
+specified in the `nodes` list. Prior to this, the admin would have had to deploy
+the local-storage-provisioner, created local PVs for each of the devices, and
+created a StorageClass to allow binding of the PVs to PVCs. At this point, the
+same workflow would be the same as the cloud use case, where you simply specify
+a count of devices, and a template with a StorageClass. Two notes here:
+
+1. The count of devices would need to match the number of existing devices to
+consume them all.
+1. The capacity for each device is irrelevant, since we will simply consume the
+entire storage device and get that capacity regardless of what is set for the
+PVC.
+
+#### Example 5: Multiple devices per OSD
+
+```yaml
+type: CephCluster
+name: cluster1
+...
+spec:
+  ...
+  storage:
+    ...
+    storageClassDeviceSet:
+    - name: cluster1-set1
+      count: 3
+      config:
+        metadataDevice: "/dev/rook/device1"
+      volumeClaimTemplates:
+      - metadata:
+          name: osd-data
+        spec:
+          resources:
+            requests:
+              storage: 1
```

Comment:
Don't we need units on the storage request such as `1Gi`?

Reply:
Nope. 1 is technically just one byte. Alternatively we should just omitted and, if not specified, use a value of `1` on the backend.

Reply:
Yeah we just won't get very far with an OSD on top of a single byte!